lib/rack/session/cookie.rb

```diff
@@ -48,55 +49,36 @@ module Session
     #
 
     class Cookie < Abstract::Persisted
-      # Encode session cookies as Base64
-      class Base64
+
+      # Marshal and encode with strict Base64
+      BASE64_MARSHALLED = Class.new do
         def encode(str)
-          [str].pack('m')
+          Base64.strict_encode64 ::Marshal.dump(str)
         end
 
         def decode(str)
-          str.unpack('m').first
-        end
-
-        # Encode session cookies as Marshaled Base64 data
-        class Marshal < Base64
-          def encode(str)
-            super(::Marshal.dump(str))
-          end
-
-          def decode(str)
-            return unless str
-            ::Marshal.load(super(str)) rescue nil
-          end
+          return unless str
+          _decode str
         end
 
-        # N.B. Unlike other encoding methods, the contained objects must be a
-        # valid JSON composite type, either a Hash or an Array.
-        class JSON < Base64
-          def encode(obj)
-            super(::JSON.dump(obj))
-          end
 
-          def decode(str)
-            return unless str
-            ::JSON.parse(super(str)) rescue nil
-          end
-        end
-
-        class ZipJSON < Base64
-          def encode(obj)
-            super(Zlib::Deflate.deflate(::JSON.dump(obj)))
-          end
-
-          def decode(str)
-            return unless str
-            ::JSON.parse(Zlib::Inflate.inflate(super(str)))
-          rescue
+        def _decode( str, meth = :strict_decode64 )
+          ::Marshal.load( Base64.send meth, str )
+        rescue ArgumentError
+          # Try non-strict Base64 before failing gracefully
+          if meth == :strict_decode64
+            meth = :decode64
+            retry
+          else
+            warn e.message
```

Comment:
undefined local variable or method `e' ?

Reply:
That is a very good point. I've fixed it but an accompanying test might be an idea.

Reply:
I tried my best but Base64.decode64 will not throw an ArgumentError under any circumstances which means the else branch of line 72 never gets hit. However, I'm loathe to remove it because it so obviously could become an infinite loop if that behaviour ever changes. Let me know what you think.

Circle-CI doesn't seem to like my force push, sorry.